tracing/tracing.go

```diff
@@ -0,0 +1,205 @@
+// Copyright 2021 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package tracing
+
+import (
+	"context"
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/pkg/errors"
+	config_util "github.com/prometheus/common/config"
+	"github.com/prometheus/common/version"
+	"go.opentelemetry.io/otel"
+	"go.opentelemetry.io/otel/exporters/otlp/otlptrace"
+	"go.opentelemetry.io/otel/exporters/otlp/otlptrace/otlptracegrpc"
+	"go.opentelemetry.io/otel/exporters/otlp/otlptrace/otlptracehttp"
+	"go.opentelemetry.io/otel/propagation"
+	"go.opentelemetry.io/otel/sdk/resource"
+	tracesdk "go.opentelemetry.io/otel/sdk/trace"
+	semconv "go.opentelemetry.io/otel/semconv/v1.7.0"
+	"go.opentelemetry.io/otel/trace"
+	"google.golang.org/grpc/credentials"
+
+	"github.com/prometheus/prometheus/config"
+)
+
+const serviceName = "prometheus"
+
+// Manager is capable of building, (re)installing and shutting down
+// the tracer provider.
+type Manager struct {
+	logger       log.Logger
+	done         chan struct{}
+	config       config.TracingConfig
+	shutdownFunc func() error
+}
+
+// NewManager creates a new tracing manager.
+func NewManager(logger log.Logger) *Manager {
+	return &Manager{
+		logger: logger,
+		done:   make(chan struct{}),
+	}
+}
+
+// Run starts the tracing manager. It registers the global text map propagator and error handler.
+// It is blocking.
+func (m *Manager) Run() {
+	otel.SetTextMapPropagator(propagation.TraceContext{})
+	otel.SetErrorHandler(otelErrHandler(func(err error) {
+		level.Error(m.logger).Log("msg", "OpenTelemetry handler returned an error", "err", err)
+	}))
+	<-m.done
+}
+
+// ApplyConfig takes care of refreshing the tracing configuration by shutting down
+// the current tracer provider (if any is registered) and installing a new one.
+func (m *Manager) ApplyConfig(cfg *config.Config) error {
+	// Update only if a config change is detected.
+	if m.config == cfg.TracingConfig {
```

Comment:
I think with a TLS config this is not true because of some private functions in the TLS configs. Is this tested?

Reply:
`TestReinstallingTracerProvider` should handle it: https://github.com/matej-g/prometheus/blob/54ec999c98f9abb7d71939d2052dcdd6e7029652/tracing/tracing_test.go#L42-L67

(Although maybe including a TLS config will change something?)

Reply:
Hm, all the `config.TLSConfig` struct fields are exported and can be compared (since its `string` and `bool` fields), so I would expect this comparison to work fine. I added a TLS config to the reinstall test case, which as @LeviHarrison should caught this in case the comparison would not work as expected.